tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,536 @@
+   * <p>Tensors can be released using the passed scope.
+   *
+   * @return the released result of {@code block}
+   */
+  public static <T extends Iterable<? extends Tensor>> T produceTensorsWithCleanup(Function<TensorScope, T> block) {
+    try (TensorScope scope = new TensorScope()) {
+      return scope.release(block.apply(scope));
+    }
+  }
+
+  /**
+   * Create a new tensor scope.  If {@code autoAttach} is false, will not automatically manage tensors.
+   *
+   * @see TensorScope
+   */
+  TensorScope() {
+    this.parent = currentScope();
+    currentScope.set(this);
+
+    if (this.parent != null) {
+      synchronized (this.parent) {
+        this.parent.children.add(this);
+      }
+    }
+  }
+
+  /**
+   * Closes this scope and its tensors, and any inner scopes.
+   */
+  @Override
+  public synchronized void close() {
+    if (closed) {
+      return;
+    }
+
+    children.forEach(TensorScope::close);
+    tensors.forEach(Tensor::close);
+
+    closed = true;
+
+    if (parent != null) {
+      parent.children.remove(this);
+    }
+
+    if (currentScope() == this) {
+      currentScope.set(this.parent);
+    }
+  }
+
+  /**
+   * Release the tensors and child scopes of this scope to it's parent, <b>without closing them</b>.
+   * <p>
+   * Semantically, calling this method makes all of the resources in this scope the parent's responsibility, as if this
+   * scope had never existed.
+   * <p>
+   * This will close this scope, but does not close any of it's resources.
+   *
+   * @throws IllegalStateException if this scope has no parent. If this happens, * the scope is not closed and no
+   * resources are released.
+   */
+  public synchronized void releaseToParent() {
+    release(true);
+  }
+
+  /**
+   * Release the tensors and child scopes of this scope to it's parent, or detach them if this scope has no parent.
+   * <p>
+   * Semantically, calling this method makes all of the resources in this scope the parent's responsibility, as if this
+   * scope had never existed.  It can be used in a method to transfer control to the caller, leaving how the resources
+   * are managed up to the caller.
+   * <p>
+   * This will close this scope, but does not close any of it's resources.
+   */
+  public synchronized void release() {
+    release(false);
+  }
+
+  /**
+   * Release the tensors and child scopes of this scope <b>without closing them</b>, to it's parent if it has one.
+   *
+   * <p><b>WARNING:</b> this method may release resources without assigning them to another scope if
+   * {@code requireParent} is false.  {@link #releaseToParent()} should be used instead wherever possible.
+   *
+   * @param requireParent Whether to require a parent scope to release resources to.
+   * @throws IllegalStateException if this scope has no parent, but {@code requireParent} is true. If this happens, the
+   * scope is not closed and no resources are released.
+   */
+  public synchronized void release(boolean requireParent) {
```

Comment:
I think the contract of this method is a bit unclear, as its behavior is quite different depending on the value of `requireParent` (if true, tensors remains managed while if false, they are free and must be freed explicitly).

I suggest to make it private so users should rely on `releaseToParent` or on `release`, which are more self-explanatory.